File: src/main/java/sawyern/cookiebot/commands/LeaderboardCommand.java
package sawyern.cookiebot.commands;

import discord4j.core.event.domain.message.MessageCreateEvent;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import sawyern.cookiebot.models.entity.Account;
import sawyern.cookiebot.exception.CookieException;
import sawyern.cookiebot.services.AccountService;
import sawyern.cookiebot.services.CookieService;
import sawyern.cookiebot.util.BotUtil;
import sawyern.cookiebot.util.MapUtil;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Component
@RequiredArgsConstructor
public class LeaderboardCommand extends MessageCreateEventBotCommand {

    private final CookieService cookieService;
    private final AccountService accountService;

    @Override
    public String getCommand() {
        return "leaderboard";
    }

    @Override
    public void execute(MessageCreateEvent event, List<String> args) throws CookieException {
        StringBuilder leaderboard = new StringBuilder("```");
        Map<String, Integer> leaderboardMap = new HashMap<>();

        List<Account> accounts = accountService.findAllAccount();

        for (Account account : accounts) {
            int cookies = cookieService.getAllCookiesForAccount(account.getDiscordId());
            leaderboardMap.put(account.getUsername(), cookies);
        }

        leaderboardMap = MapUtil.sortByValue(leaderboardMap);
        for (Map.Entry<String, Integer> entry : leaderboardMap.entrySet()) {
            leaderboard.append(entry.getKey())
                    .append(": ")
                    .append(entry.getValue())
                    .append("\n");
        }

        leaderboard.append("```");
        BotUtil.sendMessage(event, leaderboard.toString());
    }
}

File: src/main/kotlin/sawyern/cookiebot/commands/LeaderboardCommand.kt
package sawyern.cookiebot.commands

import discord4j.core.event.domain.message.MessageCreateEvent
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import sawyern.cookiebot.services.AccountService
import sawyern.cookiebot.services.BotUtilService
import sawyern.cookiebot.services.CookieService
import java.util.HashMap

@Component
class LeaderboardCommand @Autowired constructor(
        botUtilService: BotUtilService,
        private val cookieService: CookieService,
        private val accountService: AccountService
): MessageCreateEventBotCommand(botUtilService) {
    override fun getCommand(): String {
        return "leaderboard"
    }

    override fun getHelpText(): String {
        return "get seasonal ranking of cookies"
    }

    override fun execute(event: MessageCreateEvent, args: List<String>) {
        val leaderboard = StringBuilder("```")
        var leaderboardMap = HashMap<String, Int>()

        val accounts = accountService.findAll()

        for (account in accounts) {
            val cookies = cookieService.getCookiesForAccount(account.discordId)
            leaderboardMap[account.username] = cookies
        }

        var sortedLeaderboardMap = leaderboardMap
                .toList()
                .sortedBy { (key, value) -> value }
                .asReversed()
                .toMap()

        for ((key, value) in sortedLeaderboardMap) {
            leaderboard.append(key)
                    .append(": ")
                    .append(value)
                    .append("\n")
        }

        leaderboard.append("```")
        botUtilService.sendMessage(event, leaderboard.toString())
    }
}